File: ems-frontend/src/app/core/services/employee.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Employee, CreateEmployeeRequest, Page, ApiResponse } from '../../models';
import { map } from 'rxjs/operators';

@Injectable({
  providedIn: 'root',
})
export class EmployeeService {
  private apiUrl = 'http://localhost:8080/api/employees';

  constructor(private http: HttpClient) {}

  getAllEmployees(
    page: number = 0,
    size: number = 10,
    search: string = '',
  ): Observable<Page<Employee>> {
    let params = new HttpParams().set('page', page.toString()).set('size', size.toString());

    if (search) {
      params = params.set('search', search);
    }

    return this.http.get<ApiResponse<Page<Employee>>>(this.apiUrl, { params }).pipe(
      map(response => response.data)
    );
  }

  getEmployeeById(id: string): Observable<Employee> {
    return this.http.get<ApiResponse<Employee>>(`${this.apiUrl}/${id}`).pipe(
      map(response => response.data)
    );
  }

  createEmployee(request: CreateEmployeeRequest): Observable<Employee> {
    return this.http.post<ApiResponse<Employee>>(this.apiUrl, request).pipe(
      map(response => response.data)
    );
  }

  updateEmployee(id: string, request: CreateEmployeeRequest): Observable<Employee> {
    return this.http.put<ApiResponse<Employee>>(`${this.apiUrl}/${id}`, request).pipe(
      map(response => response.data)
    );
  }

  deleteEmployee(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: ems-backend/src/main/java/com/ems/modules/employee/service/EmployeeService.java
package com.ems.modules.employee.service;

import com.ems.modules.employee.dto.request.CreateEmployeeRequest;
import com.ems.modules.employee.dto.response.EmployeeResponse;

import com.ems.modules.employee.mapper.EmployeeMapper;

import com.ems.entity.Employee;
import com.ems.modules.auth.entity.User;
import com.ems.common.exception.ResourceNotFoundException;
import com.ems.common.exception.ResourceConflictException;
import com.ems.repository.EmployeeRepository;
import com.ems.modules.auth.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class EmployeeService {

    private final EmployeeRepository employeeRepository;
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final EmployeeMapper employeeMapper;

    @Transactional
    public EmployeeResponse createEmployee(CreateEmployeeRequest request) {
        // 1. Check if email exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new ResourceConflictException("Email already exists: " + request.getEmail());
        }

        // 2. Create User account
        User user = User.builder()
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .role(User.Role.EMPLOYEE)
                .build();
        user = userRepository.save(user);

        // 3. Create Employee profile
        Employee employee = employeeMapper.toEntity(request, user);
        employee = employeeRepository.save(employee);

        return employeeMapper.toResponse(employee);
    }

    @Transactional(readOnly = true)
    public Page<EmployeeResponse> getAllEmployees(String search, Pageable pageable) {
        Page<Employee> employees;
        if (search != null && !search.isEmpty()) {
            employees = employeeRepository.search(search, pageable);
        } else {
            employees = employeeRepository.findAll(pageable);
        }
        return employees.map(employeeMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public EmployeeResponse getEmployeeById(UUID id) {
        Employee employee = employeeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Employee", "id", id));
        return employeeMapper.toResponse(employee);
    }

    @Transactional
    public EmployeeResponse updateEmployee(UUID id, CreateEmployeeRequest request) {
        Employee employee = employeeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Employee", "id", id));

        // Update user email if changed (and check uniqueness)
        /*
         * // Logic to update email would go here, checking conflict
         */

        employeeMapper.updateEntity(employee, request);
        employee = employeeRepository.save(employee);

        return employeeMapper.toResponse(employee);
    }

    @Transactional
    public void deleteEmployee(UUID id) {
        Employee employee = employeeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Employee", "id", id));

        // 1. Unassign projects (Set assignedTo = null)
        // Using a custom repository method would be even better for performance, 
        // but for now, we'll iterate and rely on JPA dirty checking within the transaction.
        if (employee.getProjects() != null) {
            employee.getProjects().forEach(p -> p.setAssignedTo(null));
        }

        // 2. The User is the 'owner' of the relationship in some senses, 
        // but Employee holds the FK. 
        User user = employee.getUser();

        // 3. Delete Employee (Cascades to Attendance, Leaves, Salaries)
        employeeRepository.delete(employee);

        // 4. Finally delete the User account
        if (user != null) {
            userRepository.delete(user);
        }
    }
}
